### ISSUE TITLE
RuntimeError failed to encode SpanLink with 2.3.0

### ISSUE THREAD

**Issue by lecrepont01** [DataDog/dd-trace-py#7701]
### Summary of problem

We have experienced `RuntimeError` issues with the extension starting from this morning and the dependabot update to 2.3.0.

Stack trace :
```
ValueError: Failed to encode SpanLink. span_id=None contains an unsupported type, <class 'NoneType'>
  File "ddtrace/internal/_encoding.pyx", line 488, in ddtrace.internal._encoding.MsgpackEncoderBase._pack_trace
  File "ddtrace/internal/_encoding.pyx", line 747, in ddtrace.internal._encoding.MsgpackEncoderV03.pack_span
  File "ddtrace/internal/_encoding.pyx", line 592, in ddtrace.internal._encoding.MsgpackEncoderV03._pack_links

RuntimeError: failed to pack span: <Span(id=6671936502756880443,trace_id=14943038461850520012489475877480943891,parent_id=4075607385283888244,name=fastapi.request)>. Exception: Failed to encode SpanLink. span_id=None contains an unsupported type, <class 'NoneType'>
  File "starlette/applications.py", line 122, in __call__
    await self.middleware_stack(scope, receive, send)
  File "ddtrace/contrib/asgi/middleware.py", line 275, in __call__
    reraise(exc_type, exc_val, exc_tb)
  File "six.py", line 719, in reraise
    raise value
  File "ddtrace/contrib/asgi/middleware.py", line 270, in __call__
    return await self.app(scope, receive, wrapped_send)
  File "starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "starlette/middleware/cors.py", line 83, in __call__
    await self.app(scope, receive, send)
  File "starlette/middleware/exceptions.py", line 79, in __call__
    raise exc
  File "starlette/middleware/exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
  File "fastapi/middleware/asyncexitstack.py", line 20, in __call__
    raise e
  File "fastapi/middleware/asyncexitstack.py", line 17, in __call__
    await self.app(scope, receive, send)
  File "starlette/routing.py", line 718, in __call__
    await route.handle(scope, receive, send)
  File "starlette/routing.py", line 276, in handle
    await self.app(scope, receive, send)
  File "starlette/routing.py", line 69, in app
    await response(scope, receive, send)
  File "starlette/responses.py", line 171, in __call__
    await send({"type": "http.response.body", "body": self.body})
  File "starlette/middleware/exceptions.py", line 65, in sender
    await send(message)
  File "starlette/middleware/errors.py", line 159, in _send
    await send(message)
  File "ddtrace/contrib/asgi/middleware.py", line 239, in wrapped_send
    span.finish()
  File "ddtrace/span.py", line 272, in finish
    self._finish_ns(time_ns())
  File "ddtrace/span.py", line 285, in _finish_ns
    cb(self)
  File "ddtrace/tracer.py", line 747, in _on_span_finish
    p.on_span_finish(span)
  File "ddtrace/internal/processor/trace.py", line 251, in on_span_finish
    self._writer.write(spans)
  File "ddtrace/internal/writer/writer.py", line 324, in write
    self._write_with_client(client, spans=spans)
  File "ddtrace/internal/writer/writer.py", line 347, in _write_with_client
    client.encoder.put(spans)
  File "ddtrace/internal/_encoding.pyx", line 498, in ddtrace.internal._encoding.MsgpackEncoderBase.put
  File "ddtrace/internal/_encoding.pyx", line 502, in ddtrace.internal._encoding.MsgpackEncoderBase.put
  File "ddtrace/internal/_encoding.pyx", line 525, in ddtrace.internal._encoding.MsgpackEncoderBase.put
  File "ddtrace/internal/_encoding.pyx", line 506, in ddtrace.internal._encoding.MsgpackEncoderBase.put
  File "ddtrace/internal/_encoding.pyx", line 490, in ddtrace.internal._encoding.MsgpackEncoderBase._pack_trace
```

### Which version of dd-trace-py are you using?

2.3.0

### Which libraries and their versions are you using?

<details>
  <summary>`pip freeze`</summary>
  aiofiles==23.2.1
  alabaster==0.7.13
  alembic==1.12.1
  alembic-utils==0.8.2
  annotated-types==0.6.0
  anyio==3.7.1
  anys==0.3.0
  asgi-lifespan==2.1.0
  attrs==23.1.0
  Authlib==1.2.1
  Babel==2.13.1
  beautifulsoup4==4.12.2
  boltons==21.0.0
  bracex==2.4
  bytecode==0.15.1
  cachetools==5.3.2
  cattrs==23.2.1
  certifi==2023.11.17
  cffi==1.16.0
  charset-normalizer==3.3.2
  click==8.1.7
  click-option-group==0.5.6
  colorama==0.4.6
  coverage==7.3.2
  cryptography==41.0.5
  daiquiri==3.2.3
  datadog==0.47.0
  ddsketch==2.0.4
  ddtrace==2.2.0
  deepdiff==6.7.1
  defusedxml==0.7.1
  Deprecated==1.2.14
  docutils==0.20.1
  dulwich==0.21.6
  envier==0.4.0
  execnet==2.0.2
  face==22.0.0
  fastapi==0.104.1
  filelock==3.13.1
  first==2.0.2
  flupy==1.2.0
  glom==22.1.0
  google-api-core==2.14.0
  google-auth==2.23.4
  google-cloud-core==2.3.3
  google-cloud-storage==2.13.0
  google-crc32c==1.5.0
  google-resumable-media==2.6.0
  googleapis-common-protos==1.61.0
  graphql-utils==0.4
  greenlet==3.0.1
  gunicorn==21.2.0
  h11==0.14.0
  h2==4.1.0
  hiredis==2.2.3
  hpack==4.0.0
  httpcore==1.0.2
  httptools==0.6.1
  httpx==0.25.1
  hyperframe==6.0.1
  idna==3.4
  imagesize==1.4.1
  imia==0.5.3
  importlib-metadata==6.8.0
  iniconfig==2.0.0
  itsdangerous==2.1.2
  Jinja2==3.1.2
  jsonschema==4.20.0
  jsonschema-specifications==2023.11.1
  Mako==1.3.0
  markdown-it-py==3.0.0
  markdownify==0.11.6
  MarkupSafe==2.1.3
  msgpack==1.0.7
  multidict==6.0.4
  mypy==1.7.0
  mypy-extensions==1.0.0
  numpy==1.26.2
  opentelemetry-api==1.21.0
  ordered-set==4.1.0
  packaging==23.2
  parse==1.19.1
  passlib==1.7.4
  pastel==0.2.1
  pathspec==0.11.2
  pbr==6.0.0
  peewee==3.17.0
  pgvector==0.2.3
  pluggy==1.3.0
  poethepoet==0.24.4
  protobuf==4.25.1
  psycopg==3.1.13
  psycopg-binary==3.1.13
  pyasn1==0.5.1
  pyasn1-modules==0.3.0
  pycparser==2.21
  pydantic==2.5.1
  pydantic-settings==2.1.0
  pydantic_core==2.14.3
  pyenchant==3.2.2
  Pygments==2.17.1
  PyJWT==2.8.0
  pytest==7.4.3
  pytest-asyncio==0.21.1
  pytest-cov==4.1.0
  pytest-flakefinder==1.1.0
  pytest-github-actions-annotate-failures==0.2.0
  pytest-rerunfailures==12.0
  pytest-split==0.8.1
  pytest-timeout==2.2.0
  pytest-xdist==3.4.0
  python-dotenv==1.0.0
  python-json-logger==2.0.7
  python-multipart==0.0.6
  PyYAML==6.0.1
  redis==5.0.0
  referencing==0.31.0
  regex==2023.10.3
  reno==4.0.0
  requests==2.31.0
  respx==0.20.2
  rich==13.7.0
  rpds-py==0.13.1
  rsa==4.9
  ruamel.yaml==0.17.40
  ruamel.yaml.clib==0.2.8
  ruff==0.1.5
  semgrep==1.50.0
  sentry-sdk==1.35.0
  setproctitle==1.3.3
  six==1.16.0
  sniffio==1.3.0
  snowballstemmer==2.2.0
  soupsieve==2.5
  Sphinx==7.2.6
  sphinxcontrib-applehelp==1.0.7
  sphinxcontrib-devhelp==1.0.5
  sphinxcontrib-htmlhelp==2.0.4
  sphinxcontrib-jquery==4.1
  sphinxcontrib-jsmath==1.0.1
  sphinxcontrib-qthelp==1.0.6
  sphinxcontrib-serializinghtml==1.1.9
  sphinxcontrib-spelling==8.0.0
  SQLAlchemy==2.0.23
  SQLAlchemy-Utils==0.41.1
  starlette==0.27.0
  starsessions==2.1.3
  tenacity==8.2.3
  tomli==2.0.1
  types-first==2.0.5.2
  types-pkg-resources==0.1.3
  types-pyOpenSSL==23.3.0.0
  types-python-dateutil==2.8.19.14
  types-PyYAML==6.0.12.12
  types-redis==4.6.0.10
  typing_extensions==4.8.0
  Unidecode==1.3.7
  urllib3==1.26.18
  uvicorn==0.24.0.post1
  uvloop==0.19.0
</details>

### How can we reproduce your problem?

In that piece of code where the error started popping we have an fast API endpoint raising a 404.
raise fastapi.HTTPException(
          status_code=404,
          detail=f'item {item} does not exist',
      )

### What is the result that you get?

See stack trace

### What is the result that you expected?

Error and span reported correctly as usual.
We reverted to 2.2.0

**Comment by mabdinur** [comment]
Hi @lecrepont01,

Thanks for opening this issue. This looks a bug introduced by recent changes to our distributed tracing: https://github.com/DataDog/dd-trace-py/commit/a33ae8cbce7a889e93373b94a31d4f1409cec53d#diff-7f9cd09cc68505cb98c9e4b73fe950816115dce90379cd2db5a0616abb151b92. 

Can you share with us:
 - the ddtrace http headers received by your endpoint before this error occurs (ex: `x-datadog-*`, `traceparent`, `tracestate` headers) . 
 - the distributed tracing configurations set on your applications (ex: the value of the following environment variables if they are set  [DD_TRACE_PROPAGATION_STYLE](https://ddtrace.readthedocs.io/en/stable/configuration.html#DD_TRACE_PROPAGATION_STYLE),  [DD_TRACE_PROPAGATION_STYLE_EXTRACT](https://ddtrace.readthedocs.io/en/stable/configuration.html#DD_TRACE_PROPAGATION_STYLE_EXTRACT), [DD_TRACE_PROPAGATION_STYLE_INJECT](https://ddtrace.readthedocs.io/en/stable/configuration.html#DD_TRACE_PROPAGATION_STYLE_INJECT)).

With this information we can put together a reproduction. I suspect that span_id is being set to None here: https://github.com/DataDog/dd-trace-py/blob/v2.3.0/ddtrace/propagation/http.py#L846.

**Comment by iherasymenko** [comment]
It seems like the issue hasn't been fully fixed in 2.3.1. Just got this error.

```
StopIteration: null
  File "ddtrace/tracing/trace_handlers.py", line 55, in __next__
    return next(self.__wrapped__)
  File "werkzeug/wsgi.py", line 256, in __next__
    return self._next()

ValueError: Failed to encode SpanLink. trace_id=2537783442963587957 contains an unsupported type, <class 'int'>
  File "ddtrace/internal/_encoding.pyx", line 488, in ddtrace.internal._encoding.MsgpackEncoderBase._pack_trace
  File "ddtrace/internal/_encoding.pyx", line 747, in ddtrace.internal._encoding.MsgpackEncoderV03.pack_span
  File "ddtrace/internal/_encoding.pyx", line 592, in ddtrace.internal._encoding.MsgpackEncoderV03._pack_links

RuntimeError: failed to pack span: <Span(id=3125101130656327656,trace_id=135867149259679370681651600671496976531,parent_id=12174655768984733354,name=flask.request)>. Exception: Failed to encode SpanLink. trace_id=2537783442963587957 contains an unsupported type, <class 'int'>
  File "ddtrace/tracing/trace_handlers.py", line 57, in __next__
    self._finish_spans()
  File "ddtrace/tracing/trace_handlers.py", line 75, in _finish_spans
    self._self_parent_span.finish()
  File "ddtrace/span.py", line 272, in finish
    self._finish_ns(time_ns())
  File "ddtrace/span.py", line 285, in _finish_ns
    cb(self)
  File "ddtrace/tracer.py", line 747, in _on_span_finish
    p.on_span_finish(span)
  File "ddtrace/internal/processor/trace.py", line 251, in on_span_finish
    self._writer.write(spans)
  File "ddtrace/internal/writer/writer.py", line 324, in write
    self._write_with_client(client, spans=spans)
  File "ddtrace/internal/writer/writer.py", line 347, in _write_with_client
    client.encoder.put(spans)
  File "ddtrace/internal/_encoding.pyx", line 498, in ddtrace.internal._encoding.MsgpackEncoderBase.put
  File "ddtrace/internal/_encoding.pyx", line 502, in ddtrace.internal._encoding.MsgpackEncoderBase.put
  File "ddtrace/internal/_encoding.pyx", line 525, in ddtrace.internal._encoding.MsgpackEncoderBase.put
  File "ddtrace/internal/_encoding.pyx", line 506, in ddtrace.internal._encoding.MsgpackEncoderBase.put
  File "ddtrace/internal/_encoding.pyx", line 490, in ddtrace.internal._encoding.MsgpackEncoderBase._pack_trace
```

```
client	

<ddtrace.internal.writer.writer_client.AgentWriterClientV4 object at 0x7f64f5940ed0>

self	

AgentWriter(status=<ServiceStatus.RUNNING: 'running'>, _interval=1.0)

spans	

[<Span(id=1677576634454439942,trace_id=85660492545568585082154361004331002392,parent_id=1722468387988178087,name=flask.request)>, <Span(id=2900904841436811886,trace_id=85660492545568585082154361004331002392,parent_id=1677576634454439942,name=flask.application)>, <Span(id=7060135266343858286,trace_id=85660492545568585082154361004331002392,parent_id=2900904841436811886,name=sentry_sdk.integrations.flask._request_started)>, <Span(id=1828519013878440863,trace_id=85660492545568585082154361004331002392,parent_id=2900904841436811886,name=flask.preprocess_request)>, <Span(id=3973354597771674484,trace_id=85660492545568585082154361004331002392,parent_id=1828519013878440863,name=flask_limiter.extension._check_request_limit)>, <Span(id=13018030091899534568,trace_id=85660492545568585082154361004331002392,parent_id=1828519013878440863,name=app.before_request)>, <Span(id=5508624154637222398,trace_id=85660492545568585082154361004331002392,parent_id=2900904841436811886,name=flask.dispatch_request)>, <Span(id=4751179752361833950,trace_id=85660492545568585082154361004331002392,parent_id=5508624154637222398,name=clients.read_client)>, <Span(id=6654300785753861651,trace_id=85660492545568585082154361004331002392,parent_id=4751179752361833950,name=pymongo.checkout)>, <Span(id=7148787690478966493,trace_id=85660492545568585082154361004331002392,parent_id=6654300785753861651,name=pymongo.cmd)>]
```

**Comment by ZStriker19** [comment]
Hi @iherasymenko, thanks for reaching out! Interesting error and a bit more troublesome to guess the root cause. We're hitting [the exception here](https://github.com/DataDog/dd-trace-py/blob/main/ddtrace/internal/_encoding.pyx#L575-L592). 

A time when an `int` isn't a `Number`, or something else is happening. Would you mind telling us the following so we can try and recreate?
1. What os/version are you seeing this error on?
2. What python version are you on? (I tried 2.7 and 3.12, but no difference)
3. If you run 
```
from numbers import Number
isinstance(2537783442963587957, Number)
```
On the same platform/python version, what do you get?

4. Could you run with DD_TRACE_DEBUG=true (just to startup your app and send us the output) so we can get startup information?

**Comment by iherasymenko** [comment]
@ZStriker19 sorry for the late reply.

We use the Docker base image `python:3.12-bookworm-slim` which is run on an amd64 cloud machine. `ddtrace` version is 2.3.2.

The code above yields `True`.

Looks like the issue is only happening when the Flask endpoint is hit via a DataDog Synthetics test. I've been able to _reliably_ reproduce the issue in a dev environment. 

https://app.datadoghq.com/synthetics/create

![image](https://github.com/DataDog/dd-trace-py/assets/2904801/87d236e0-eba1-4c6b-bc5c-68eb83f71a55)


The request itself doesn't fail; however, the logs contain the following error.


<img width="1004" alt="image" src="https://github.com/DataDog/dd-trace-py/assets/2904801/a69c9915-b1a7-4d77-a5ba-af344e6b2180">

I'll try to create a minimal reproducer for you.

**Comment by iherasymenko** [comment]
Just wanted to confirm that the issue is fixed in 2.4.0. Thanks @mabdinur !

**Comment by lecrepont01** [comment]
Same here. Thanks.
